File: src/c#/philosopher_eat_for_many_turns.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;

namespace Philosophers
{
    public enum StatesPhilosopher
    {
        thinking = 0,
        hungry = 1,
        eating = 2,
        finished_eat = 3
    }

    public class Philosopher
    {
        private Semaphore[] forks;
        private Thread[] philosopherThread;
        private StatesPhilosopher[] statesPhilosopher;

        public int CountPhilosopher { get; private set; }

        public Philosopher(int countPhilosopher = 5)
        {
            forks = new Semaphore[countPhilosopher];
            philosopherThread = new Thread[countPhilosopher];
            statesPhilosopher = new StatesPhilosopher[countPhilosopher];
            CountPhilosopher = countPhilosopher;

            for (int i = 0; i < countPhilosopher; i++)
            {
                forks[i] = new Semaphore(1, 1);
                statesPhilosopher[i] = StatesPhilosopher.thinking;
            }
        }

        private void TakeForks(int id_philosopher)
        {
            forks[id_philosopher].WaitOne();
            forks[(id_philosopher + 1) % CountPhilosopher].WaitOne();
        }

        private void ReleaseForks(int id_philosopher)
        {
            forks[id_philosopher].Release();
            forks[(id_philosopher + 1) % CountPhilosopher].Release();
        }

        private void Eat(int id_philosopher)
        {
            Random random = new Random();

            while (true)
            {
                if (random.Next(0, 2) == 0)
                {
                    statesPhilosopher[id_philosopher] = StatesPhilosopher.thinking;
                    Console.WriteLine($"[*] Filosofo {id_philosopher} pensando.");
                    Thread.Sleep(TimeSpan.FromSeconds(random.Next(2, 5)));
                }
                else
                {
                    statesPhilosopher[id_philosopher] = StatesPhilosopher.hungry;
                    Console.WriteLine($"[@] Filosofo {id_philosopher} hambriento.");

                    TakeForks(id_philosopher);

                    statesPhilosopher[id_philosopher] = StatesPhilosopher.eating;
                    Console.WriteLine($"[+] Filosofo {id_philosopher} comiendo.");
                    Thread.Sleep(TimeSpan.FromSeconds(random.Next(3, 5)));
                    statesPhilosopher[id_philosopher] = StatesPhilosopher.finished_eat;
                    Console.WriteLine($"[!] Filosofo {id_philosopher} termino de comer.");

                    ReleaseForks(id_philosopher);
                }

                Thread.Sleep(TimeSpan.FromSeconds(1));
            }
        }

        public void Run()
        {
            for (int i = 0; i < philosopherThread.Length; i++)
            {
                int idx = i;
                philosopherThread[idx] = new Thread(() => Eat(idx));
                philosopherThread[idx].Start();
            }

            for (int i = 0; i < philosopherThread.Length; i++)
            {
                philosopherThread[i].Join();
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Philosopher forManyTurn = new Philosopher(5);
            forManyTurn.Run();
        }
    }
}

File: src/python/philosopher_eat_for_many_turns.py
import enum
import time
import random
import threading

class StatesPhilosopher(enum.Enum):
    '''
    Representa los posibles estados de los filosofos.
    '''
    thinking = 1
    hungry = 2
    eating = 3
    finished_eat = 4

class Philosopher:

    def __init__(self, count_philosopher=5):
        self.count_philosopher = count_philosopher
        self.forks = [ threading.Semaphore(1) 
                                    for _ in range(self.count_philosopher) ]
        self.state_philosopher = [ StatesPhilosopher.thinking 
                                    for _ in range(self.count_philosopher) ]
        self.start_threads()

    def take_forks(self, id_philosopher):
        '''
        Intenta tomar los tenedores adyacentes 
        al filosofo especificado.
        '''
        self.forks[id_philosopher].acquire()
        self.forks[(id_philosopher + 1) % self.count_philosopher].acquire()

    def release_forks(self, id_philosopher):
        '''
        Libera los tenedores adyacentes del filosofo especificado.
        '''
        self.forks[id_philosopher].release()
        self.forks[(id_philosopher + 1) % self.count_philosopher].release()

    def eat(self, id_philosopher):
        '''
        Funcion que va ejecutar cada filosofo o hilo.
        '''
        while True:

            if random.randint(0, 2) == 0:
                self.state_philosopher[id_philosopher] = StatesPhilosopher.thinking
                print(f'[*] Filosofo {id_philosopher} pensando.')
                time.sleep(random.randint(2, 5))
            else:
                self.state_philosopher[id_philosopher] = StatesPhilosopher.hungry
                print(f'[@] Filosofo {id_philosopher} hambiento.')
                
                self.take_forks(id_philosopher)

                self.state_philosopher[id_philosopher] = StatesPhilosopher.eating
                print(f'[+] Filosofo {id_philosopher} comiendo.')
                time.sleep(random.randint(3, 5))
                self.state_philosopher[id_philosopher] = StatesPhilosopher.finished_eat
                print(f'[!] Filosofo {id_philosopher} termino de comer.')

                self.release_forks(id_philosopher)

            time.sleep(1)

    def start_threads(self):
        '''
        Comienza a correr los hilos correspondiente 
        a cada filosofo.
        '''
        threads = [ threading.Thread(target=self.eat, args=(id,)) 
                        for id in range(self.count_philosopher) ]

        for thread in threads:
            thread.start()

        for thread in threads:
            thread.join()

def main():
    
    Philosopher()

if __name__ == '__main__':
    main()